File: leetcode/445.py
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        x1, x2 = 0, 0
        while l1:
            x1 = x1 * 10 + l1.val
            l1 = l1.next
        while l2:
            x2 = x2 * 10 + l2.val
            l2 = l2.next
        x = x1 + x2
        head = ListNode(0)
        if x == 0: return head
        while x:
            v, x = x % 10, x / 10
            head.next, head.next.next = ListNode(v), head.next
        return head.next

File: leetcode/445.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#define pb push_back
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        vector<int> a, b, c;
        while (l1) {
            a.pb(l1 -> val);
            l1 = l1 -> next;
        }
        while (l2) {
            b.pb(l2 -> val);
            l2 = l2 -> next;
        }
        int i = a.size() - 1, j = b.size() - 1;
        int carry = 0;
        ListNode* head = NULL, *now = NULL;
        while (i >= 0 || j >= 0 || carry > 0) {
            int sum = carry;
            if (i >= 0) sum += a[i--];
            if (j >= 0) sum += b[j--];
            carry = sum / 10;
            now = new ListNode(sum % 10);
            now -> next = head;
            head = now;
        }
        return head;
    }
};